rclcpp_action/include/rclcpp_action/client_goal_handle.hpp

```diff
@@ -0,0 +1,64 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP_ACTION__CLIENT_GOAL_HANDLE_HPP_
+#define RCLCPP_ACTION__CLIENT_GOAL_HANDLE_HPP_
+
+#include <rcl_action/action_client.h>
+
+#include <functional>
+#include <memory>
+
+#include "rclcpp_action/visibility_control.hpp"
+
+namespace rclcpp_action
+{
+// Forward declarations
+template<typename ACTION>
+class Client;
+
+template<typename ACTION>
+class ClientGoalHandle
+{
+public:
+  RCLCPP_ACTION_PUBLIC
+  virtual ~ClientGoalHandle();
+
+  /// TODO(sloretz) examples have this on client as `async_cancel_goal(goal_handle)`
+  std::future<bool>
+  async_cancel();
+
+  // TODO(sloretz) `async_result` to make method names consistent?
+  std::future<typename ACTION::Result>
+  result_future();
+
+private:
+  // The templated Server creates goal handles
+  friend Client<ACTION>;
```

Comment:
@sloretz I don't know if this is a typical pattern in ROS2, but why not using an abstract base class and have the client implement their own?

Reply:
I don't using `friend` and a private constructor is a typical pattern. The base class approach would work too.

Reply:
@hidmic is 193eb4a8bd8010f7ed5f5429e8c4d1a375fc759d what you had in mind?

Reply:
Hmm, I was actually thinking of something like (in client_goal_handle.hpp):

```c++
class ClientGoalHandle {
  public:
    virtual std::future<bool> async_cancel() = 0;
};
```

and then (in `client.cpp`):
```c++
class SimpleClientGoalHandle : public ClientGoalHandle {
   public:
      std::future<bool> async_cancel() override {
         ...
      }
};
```
External code never instantiates a `*GoalHandle` because it knows nothing of it. We have to watch out for slicing though.

Reply:
I don't think `ClientGoalHandleImpl` can be in `client.cpp` because `std::future<typename ACTION::Result> async_result()` requires knowing the user's type.

Reply:
Argh, right, there's no common base type for action user defined types, silly me. 

I think it's fine. Another alternative (for the sake of completeness :) ) would be to define the goal handle subclass as a private nested class of the client/server class. Same code, no `friend`ship.